src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/StreamPipeReader.cs

```diff
@@ -216,27 +217,88 @@ public override async ValueTask<ReadResult> ReadAsync(CancellationToken cancella
                 var isCanceled = false;
                 try
                 {
-                    // This optimization only makes sense if we don't have anything buffered
-                    if (UseZeroByteReads && _bufferedBytes == 0)
+                    await ReadAsyncInternal(tokenSource.Token).ConfigureAwait(false);
```

Comment:
You added an extra state machine. Can you undo this.

Reply:
Done. Do we have another solution here to avoid code duplication?